### DISCUSSION TITLE
How can I control a trace so that only the root span is called?

### DISCUSSION THREAD

**Discussion by zerocooldog** [open-telemetry/opentelemetry-java-instrumentation#12760] [Q&A]
Looking into the OTEL instrumentation module, I found that there are advice classes related to servlets.

```
https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/servlet
```

And by registering a CustomSpanProcessor, I get that the context object in the onStart method of the SpanProcessor logs information such as HttpRouteState or opentelemetry-servlet-app-server-bridge.

-CustomSpanProcc

```
public class Hamonica2SpanProcessor implements SpanProcessor {
    private final Logger log = Logger.getLogger(getClass().getName());
    private static final ContextKey<HttpRouteState> KEY =
            ContextKey.named("opentelemetry-http-server-route-key");
    @Override
    public void onStart(Context context, ReadWriteSpan readWriteSpan) {

        HttpRouteState httpRouteState = context.get(KEY);
        log.log(Level.SEVERE,"###### start context : {0}", new Object[]{context});
        log.log(Level.SEVERE,"###### start httpRouteState  : {0}", new Object[]{httpRouteState});
    }

...
...
}

```

- context Object toString
```
 {opentelemetry-http-server-route-key=io.opentelemetry.javaagent.shaded.instrumentation.api.internal.HttpRouteState@2e184ef5, opentelemetry-servlet-app-server-bridge=io.opentelemetry.javaagent.bootstrap.servlet.AppServerBridge@20dfb546}
```

Here are my questions:

1. I'm trying to access the HttpRouteState value from the context within a SpanProcessor, but it's always null. Am I using the context.get(KEY) method correctly? When I log the entire context object, it seems to contain the desired information.

2. I'm trying to optimize my tracing configuration. Can I initiate and terminate the root span just once, regardless of the number of child spans within a trace? I aim to reduce overhead and improve performance.

**Comment by laurit** [comment]
1. Context keys use identity comparison, you must use the exact same key instance that was used to set the value. Also route is often set after span is created so it will not be available in the `onStart` method.
2. I don't understand the question.

**Reply by zerocooldog** [reply]
@laurit 1. OK

2. Let me ask you again.

When implementing a SpanProcessor, both root span and child spans are invoked multiple times. However, I only want to call root span. Is there a way to achieve this?

Currently, I’m temporarily using the following function to identify root spans:
```
boolean isLocalServerRoot(ReadableSpan readableSpan) {
    return SpanKind.SERVER == readableSpan.getKind() 
        && (!readableSpan.getParentSpanContext().isValid() 
        || readableSpan.getParentSpanContext().isRemote());
}

```

**Reply by laurit** [reply]
I'm still not quite sure what you want to do. If you want to reduce the number of spans you could disable the instrumentations that produce unwanted spans. You could also try using a sampler or look into https://github.com/open-telemetry/opentelemetry-java/pull/6375 that allows disabling tracer for specified instrumentation scope.

**Reply by zerocooldog** [reply]
@laurit I want to ensure that the first span is processed only once, and from the second span onward, they should not be processed by the SpanProcessor. The first span is a server span (WAS or servlet).

If you don’t understand my question, I’ll look it up myself.
My English skills are not very good, so it’s difficult for me to write long, detailed questions. Thank you.
